File: model/src/main/java/io/quartic/geojson/FeatureCollection.java
package io.quartic.geojson;

import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.fasterxml.jackson.annotation.JsonTypeName;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import io.quartic.common.SweetStyle;
import org.immutables.value.Value;

import java.util.List;
import java.util.Map;
import java.util.Optional;

import static com.fasterxml.jackson.annotation.JsonTypeInfo.As;
import static com.fasterxml.jackson.annotation.JsonTypeInfo.Id;

@SweetStyle
@Value.Immutable
@JsonTypeName("FeatureCollection")
@JsonSerialize(as = FeatureCollectionImpl.class)
@JsonDeserialize(as = FeatureCollectionImpl.class)
@JsonTypeInfo(use= Id.NAME, include= As.PROPERTY, property="type", defaultImpl = FeatureCollectionImpl.class)
public interface FeatureCollection {
    List<Feature> features();

    @Value.Parameter(false)
    Optional<Map<String, Object>> crs();
}

File: common-core/src/main/kotlin/io/quartic/common/geojson/FeatureCollection.kt
package io.quartic.common.geojson

import com.fasterxml.jackson.annotation.JsonTypeName

@JsonTypeName("FeatureCollection")
data class FeatureCollection @JvmOverloads constructor(
        val features: List<Feature>,
        // TODO
        val crs: Map<String, Any>? = emptyMap(),
        val bbox: List<Double>? = emptyList()
) : GeoJsonObject
